temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingSpanUtils.java

```diff
@@ -0,0 +1,150 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.opentracing.internal;
+
+import com.google.common.base.MoreObjects;
+import io.opentracing.References;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+public class OpenTracingSpanUtils {
+  public static Tracer.SpanBuilder createWorkflowStartSpan(
```

Comment:
Not exactly. Each of these methods, other than just tags, encodes two things - a reference type and if explicit parent span is expected/allowed in this scenario.
I would prefer to leave them the current way and it was my intention to make these methods in such a rigid and specific way, so each scenario is super clean - what are the expected tags, if we expect explicitly passed parent span, and what is the reference type. And we have a single place where we can see the difference between these scenarios.

Reply:
I could see the benefit if these functions were called multiple times, but they are called only once each, which only unnecessarily complicates the code by adding another level of indirection.

Reply:
I think that this is a good way. It's not a level of indirection, it's a level of incapsulation. Some permutations of parameters are incorrect and these methods provide the required incapsulation to ensure the correctness of permutations of parameter values.

Reply:
I like @Spikhalskiy  approach more. Mostly because all the ugliness of creating tag hashmaps is encapsulated in one place. I think this is about the style which is very subjective. I would leave it as it is just to avoid bikeshedding.

Reply:
Made it a non-static factory and hid the logic of span operation name creation inside to give an easy entry point for providing extensible strategies in the future if needed.